scripts/build-metadata.js

```diff
@@ -0,0 +1,53 @@
+const fs = require("fs");
+const yaml = require("js-yaml");
+const glob = require("glob");
+
+// Load architecture metadata.
+const architectures = yaml.load(fs.readFileSync("metadata/architectures.yaml", "utf-8"));
+
+// Load template-group metadata.
+const templateGroups = glob.sync("metadata/groups/*.yaml")
+    .map(groupYamlFile => {
+        return yaml.load(fs.readFileSync(groupYamlFile, "utf-8"));
+    });
+
+// Load all templates.
+const allTemplates = glob.sync("*/Pulumi.yaml")
+
+    // For each one, parse it into JSON and change the name and description.
+    .map(pulumiYamlFile => {
+        const template = yaml.load(fs.readFileSync(pulumiYamlFile, "utf-8"));
+        const templateName = pulumiYamlFile.replace("/Pulumi.yaml", "");
+        return {
+            name: templateName,
+            runtime: template.runtime,
+            template: template.template,
+        };
+    })
+    .reduce((templates, template) => {
+        templates[template.name] = template;
+        return templates;
+    }, {});
+
+// Build the list of architectures and groups.
+const architectureTemplates = Object.keys(architectures)
+    .map(arch => architectures[arch])
+    .map(arch => {
+        return {
+            ...arch,
+            ...{
+                groups: templateGroups.filter(group => {
+                    return group.kind === "architecture" && group.parent === arch.slug;
+                }),
+            },
+        };
+    });
+
+// Assemble the list.
+const result = {
+    architectures: architectureTemplates,
+    templates: allTemplates,
+};
+
+// Write the list to an output file.
+fs.writeFileSync("metadata.json", JSON.stringify(result, null, 4), "utf-8");
```

Comment:
Any chance we could have some sort of check to validate the generated metadata hasn't structurally drifted. I see you added a test stage to the workflow, but not sure if it validates this. If not, or if it would be too much work, would it be possible to just have some sort of check where people need to add a label to verify they didn't touch this file when opening a PR against this repo, similar to what we do in the service repo with the changelog file. Not sure if you have any other ideas around this.

Reply:
The tests (which run on PR and push events) do verify structure, yeah, and we won't update the release unless those tests pass.

>would it be possible to just have some sort of check where people need to add a label to verify they didn't touch this file

I'm not sure I totally understand -- maybe you could explain what the concern is, and we can start from there?

Reply:
ahh nvm. yeah I see the tests now. ok that's all I wanted to confirm it validates the structure. thank you!!